### PR TITLE
Gunmods can modify a gun's  "ammo to fire"

### PR DESCRIPTION
**Pull request by John-Candlebury**
#### Summary
Mods "Gunmods can modify a gun's  "ammo to fire""

#### Purpose of change

Ensure that the vanilla laser weapon gun mods function properly within Aftershock

#### Testing

Spawn a laser gun with aftershock, install high-density capacitor mod, notice that ammo consumption has doubled.

### CODE REVIEW

File: data/mods/Aftershock/items/gunmods/laser_gunmods.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    "name": { "str": "effective emitter" },
+    "description": "A set of electronics to optimize emitter workcycle and increase overall energy efficiency.  Decreases power consumption.",
+    "ammo_to_fire_multiplier": 0.9
+  },
+  {
+    "id": "high_density_capacitor",
+    "type": "GUNMOD",
+    "copy-from": "high_density_capacitor",
+    "name": { "str": "high density capacitor" },
+    "description": "A capacitor with a higher energy density increases range and damage; at the cost of a markedly increased power consumption.",
```
Comment (actual-nh) [inline data/mods/Aftershock/items/gunmods/laser_gunmods.json:25]:
```suggestion
    "description": "A capacitor with a higher energy density increases range and damage, at the cost of doubling power consumption.",
```

File: data/mods/Aftershock/items/gunmods/laser_gunmods.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
+[
+  {
+    "id": "electrolaser_conversion",
+    "type": "GUNMOD",
+    "copy-from": "electrolaser_conversion",
+    "name": { "str": "electrolaser conversion" },
+    "description": "A set of high-tech electronics and optics.  These converts a laser pistol into a less-lethal electrolaser capable of stunning targets, at the cost of a decreased damage output and increased power consumption.",
```
Comment (actual-nh) [inline data/mods/Aftershock/items/gunmods/laser_gunmods.json:7]:
```suggestion
    "description": "A set of high-tech electronics and optics.  These convert a laser pistol into a less-lethal electrolaser capable of stunning targets, at the cost of decreased damage output and increased power consumption.",
```
Also, where does the stunning come in?

**Review by actual-nh — COMMENTED** [review]
Nice!

File: data/mods/Aftershock/items/gunmods/laser_gunmods.json (hunk 2 above)
Comment (John-Candlebury) [reply data/mods/Aftershock/items/gunmods/laser_gunmods.json]: The stunning properties are defined on the main definition in vanilla's json, this one is really just handling compatibility.

**Comment by Shodan14** [comment]
Will this finally allow the shotgun barrel adapters?

File: data/mods/Aftershock/items/gunmods/laser_gunmods.json (hunk 2 above)
Comment (ToxiClay) [reply data/mods/Aftershock/items/gunmods/laser_gunmods.json]: How do I suggest a change on a suggested change? `This converts a laser pistol...` is grammatically correct, as the singular article 'this' refers to the singular noun 'set of X.'

**Comment by Maleclypse** [comment]
> Will this finally allow the shotgun barrel adapters?

I don't believe it will allow barrel adapters that change the type of ammo fired.  It will allow a gunmod to change the amount of ammo fired.  Like a more json versatile version of the fully automatic gunmods.

**Comment by Shodan14** [comment]
> 
> 
> > Will this finally allow the shotgun barrel adapters?
> 
> I don't believe it will allow barrel adapters that change the type of ammo fired. It will allow a gunmod to change the amount of ammo fired. Like a more json versatile version of the fully automatic gunmods.

Ok :( 

Maybe one day..

**Comment by actual-nh** [comment]
"ammo_to_fire" is somewhat ambiguous as a description; it could be either amount of ammo (correct) or what ammo to use (incorrect).

**Comment by ToxiClay** [comment]
> "ammo_to_fire" is somewhat ambiguous as a description; it could be either amount of ammo (correct) or what ammo to use (incorrect).

In context, though, it would seem to be simple, because it comes up alongside a number, right?

In principle I agree with you, though; as much as possible, variables should clearly identify their purpose by their name.

**Comment by actual-nh** [comment]
> In context, though, it would seem to be simple, because it comes up alongside a number, right?

In code, yes; in the title of a PR, well...
 
> In principle I agree with you, though; as much as possible, variables should clearly identify their purpose by their name.

I may go a bit too far at times in my coding w/regard to that, with resulting very long variable names...

**Comment by Mom-Bun** [comment]
> "ammo_to_fire" is somewhat ambiguous as a description; it could be either amount of ammo (correct) or what ammo to use (incorrect).

When I added this json, that was the simplest name that made sense at the time, and what Candlebury suggested; something like "amount_to_fire" would've been too wordy I feel

File: data/mods/Aftershock/items/gunmods/laser_gunmods.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
+[
+  {
+    "id": "electrolaser_conversion",
```
Comment (Mom-Bun) [inline data/mods/Aftershock/items/gunmods/laser_gunmods.json:3]: What would be the reason for a mod that decreases damage but increases power consumption?

**Review by Mom-Bun — APPROVED** [review]
Don't have much to pick at this but I like it, it looks good! Thank you!

File: data/mods/Aftershock/items/gunmods/laser_gunmods.json (hunk 3 above)
Comment (John-Candlebury) [reply data/mods/Aftershock/items/gunmods/laser_gunmods.json:3]: Steep efficiency loss from making the gun behave in a way different than what its originally designed for.  (Plus thats how it works in vanilla too. )

**Comment by actual-nh** [comment]
> > "ammo_to_fire" is somewhat ambiguous as a description; it could be either amount of ammo (correct) or what ammo to use (incorrect).
> 
> When I added this json, that was the simplest name that made sense at the time, and what Candlebury suggested; something like "amount_to_fire" would've been too wordy I feel

No problem; under most in-use conditions, as ToxiClay wrote, it's not ambiguous.

File: data/mods/Aftershock/items/gunmods/laser_gunmods.json (hunk 3 above)
Comment (Mom-Bun) [reply data/mods/Aftershock/items/gunmods/laser_gunmods.json:3]: no no, I meant why gameplay wise WHY you'd want to have a weapon that costs more to fire, but does less damage?

Comment (John-Candlebury) [reply data/mods/Aftershock/items/gunmods/laser_gunmods.json:3]: Ah it turns the laser into a ranged stunner.  Which has some niche combat uses, the most important one being that it lets you escape things you wouldn't be able to kill using the unmodded gun.

E.X. The Av-22 will be useless against a Spartan, but with this mod you can more easily escape by stunning it

Comment (Mom-Bun) [reply data/mods/Aftershock/items/gunmods/laser_gunmods.json:3]: Ah that makes sense; thank you!
